File: Leetcode/LongestStringChain.cpp
// https://leetcode.com/problems/longest-string-chain

class Solution {
   public:
    bool check(const string& from, const string& to) {
        for (int i = 0; i < to.length(); ++i) {
            if (from == to.substr(0, i) + to.substr(i + 1)) {
                return true;
            }
        }
        return false;
    }

    int longestStrChain(std::vector<std::string>& words) {
        int n = words.size();

        vector<vector<int>> grid(n);

        sort(words.begin(), words.end(), [](const string& a, const string& b) {
            return a.length() < b.length();
        });

        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                const string& from = words[i];
                const string& to = words[j];

                if (to.length() - from.length() > 1) {
                    break;
                }

                if (to.length() - from.length() == 1 && check(from, to)) {
                    grid[i].push_back(j);
                }
            }
        }

        vector<int> dp(n, 0);
        int ans = 0;

        for (int i = 0; i < n; ++i) {
            for (int to : grid[i]) {
                dp[to] = max(dp[to], dp[i] + 1);
                ans = max(ans, dp[to]);
            }
        }

        return ans + 1;
    }
};

File: Leetcode/LongestStringChain.rs
// https://leetcode.com/problems/longest-string-chain

struct Solution {}

impl Solution {
    fn check(from: &str, to: &str) -> bool {
        let from = from.to_string();
        (0..to.len()).any(|p| from.to_owned() == format!("{}{}", &to[0..p], &to[(p + 1)..]))
    }

    pub fn longest_str_chain(mut words: Vec<String>) -> i32 {
        let n = words.len();
        let mut grid = vec![vec![]; n];

        words.sort_by(|a, b| a.len().cmp(&b.len()));

        for i in 0..n {
            for j in (i + 1)..n {
                let from = &words[i];
                let to = &words[j];

                if to.len() - from.len() > 1 {
                    break;
                }

                if to.len() - from.len() == 1 && Solution::check(from, to) {
                    grid[i].push(j);
                }
            }
        }

        let mut dp = vec![0; n];
        let mut ans = 0;

        for i in 0..n {
            for to in &grid[i] {
                let to = *to;
                dp[to] = dp[to].max(dp[i] + 1);
                ans = ans.max(dp[to]);
            }
        }

        ans + 1
    }
}
